pytheus/mmap.py

```diff
@@ -0,0 +1,344 @@
+import abc
+import collections
+import mmap
+import os
+import struct
+import typing
+
+# See https://docs.python.org/3/library/struct.html#format-characters
+StructFormatString = typing.Literal["i", "d", "?", "s"]
+StructFormatType = typing.Union[int, float, bool, str]
+
+STRUCT_INT_FORMAT: StructFormatString = "i"
+STRUCT_FLOAT_FORMAT: StructFormatString = "d"
+STRUCT_BOOL_FORMAT: StructFormatString = "?"
+STRUCT_STR_FORMAT: StructFormatString = "s"
+
+BYTE_ALIGNMENT = 4
+
+DEFAULT_BACKEND = "DEFAULT_BACKEND"
+
+
+class MmapDictTypeBackend(abc.ABC):
+    def __init__(self, mmap: mmap.mmap) -> None:
+        self._mmap = mmap
+
+    def move_right_and_pad(self, position: int, count: int):
+        """ Move a section of the array to the right and pad zeros in the gap """
+        self._mmap[position:] = b"\0" * count + self._mmap[position:-count]
+
+    def move_left_and_pad(self, position: int, count: int):
+        """ Move a section of the array to the left and pad zeros at the end """
+        self._mmap[position:] = self._mmap[position+count:] + b"\0" * count
+
+    @abc.abstractmethod
+    def get_(self, position: int) -> StructFormatType:
+        ...
+
+    @abc.abstractmethod
+    def set_(self, position: int, value: StructFormatType) -> None:
+        ...
+
+    @abc.abstractmethod
+    def size_(self, value: StructFormatType) -> int:
+        ...
+
+
+class FixedSizeMmapDictTypeBackend(MmapDictTypeBackend):
```

Comment:
This implementation works for fixed size types like integers, floats, bools, etc